internal/pkg/cli/app_deploy.go

```diff
@@ -0,0 +1,423 @@
+	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/docker"
+	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/manifest"
+	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/project"
+	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/term/prompt"
+	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/workspace"
+)
+
+// BuildAppDeployCommand builds the `app deploy` subcommand.
+func BuildAppDeployCommand() *cobra.Command {
+	input := &appDeployInput{
+		prompt: prompt.New(),
+	}
+
+	cmd := &cobra.Command{
+		Use:  "deploy",
+		Long: `Deploy an application to an environment.`,
+		Example: `
+  Deploy an application named "frontend" to a "test" environment.
+  $ archer app deploy --name frontend --env test`,
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			if err := input.init(); err != nil {
+				return err
+			}
+
+			if err := input.sourceInputs(); err != nil {
+				return err
+			}
+
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if err := input.deployApp(); err != nil {
+				return err
+			}
+
+			return nil
+		},
+		// PostRunE: func(cmd *cobra.Command, args []string) error {
+		// TODO: recommended actions?
+		// },
+	}
+
+	// TODO: make "project" a shared flag either at the `app` subcommand level or `root`?
+	cmd.Flags().StringVarP(&input.project, "project", "p", "", "project name")
+	cmd.Flags().StringVarP(&input.app, "app", "a", "", "application name")
+	cmd.Flags().StringVarP(&input.env, "env", "e", "", "environment name")
+	cmd.Flags().StringVarP(&input.imageTag, "tag", "t", "", "image tag")
+
+	return cmd
+}
+
+type appDeployInput struct {
+	project  string
+	app      string
+	env      string
+	imageTag string
+
+	projectService   projectService
+	prompt           prompter
+	ecrService       ecr.Service
+	workspaceService *workspace.Workspace
+
+	projectApplications []*archer.Application
+	projectEnvironments []*archer.Environment
+}
+
+func (i appDeployInput) String() string {
+	return fmt.Sprintf("project: %s, app: %s, env: %s, tag: %s", i.project, i.app, i.env, i.imageTag)
+}
+
+type projectService interface {
+	GetProjects() ([]*archer.Project, error)
+	GetApplications(projectName string) ([]*archer.Application, error)
+	GetEnvironments(projectName string) ([]*archer.Environment, error)
+}
+
+func (i *appDeployInput) init() error {
+	projectService, err := project.NewService()
+	if err != nil {
+		return fmt.Errorf("create project service: %w", err)
+	}
+	i.projectService = projectService
+
+	toolsAccountSession, err := session.Default()
+	if err != nil {
+		return fmt.Errorf("initialize tools account session: %w", err)
+	}
+	i.ecrService = ecr.New(toolsAccountSession)
+
+	workspaceService, err := workspace.New()
+	if err != nil {
+		return fmt.Errorf("intialize workspace service: %w", err)
+	}
+	i.workspaceService = workspaceService
+
+	return nil
+}
+
+func (i *appDeployInput) sourceInputs() error {
+	if err := i.sourceProjectName(); err != nil {
```

Comment:
Should we just insist that they be in an existing workspace and grab the project from the ws summary?

Reply:
So just error out if a project name cannot be derived?

Reply:
Yea, with a "This command must be run from within a workspace"